amethyst_tiles/src/map.rs

```diff
@@ -0,0 +1,365 @@
+    fn transform(&self) -> &Matrix4<f32>;
+
+    /// Call the underlying coordinate encoder for this map instance, which should always reduce to a u32 integer.
+    fn encode(&self, coord: &Point3<u32>) -> Option<u32>;
+
+    /// Call the underlying coordinate encoder for this map instance, which should always reduce to a u32 integer.
+    fn encode_raw(&self, coord: &(u32, u32, u32)) -> Option<u32>;
+
+    /// Call the underlying coordinate decoder for this map instance, which should always convert a u32 to a tile-space
+    /// coordinate.
+    fn decode(&self, morton: u32) -> Option<Point3<u32>>;
+
+    /// Call the underlying coordinate decoder for this map instance, which should always convert a u32 to a tile-space
+    /// coordinate.
+    fn decode_raw(&self, morton: u32) -> Option<(u32, u32, u32)>;
+}
+
+/// Generic trait over the underlying storage of a given `Map` type.
+pub trait MapStorage<T: Tile> {
+    /// Try to get the `Tile` type present at the provided coordinates.
+    fn get(&self, coord: &Point3<u32>) -> Option<&T>;
+
+    /// Try to get a mutable reference to `Tile` type present at the provided coordinates.
+    fn get_mut(&mut self, coord: &Point3<u32>) -> Option<&mut T>;
+
+    /// Try to get a mutable reference to `Tile` type present at the provided coordinates.
+    fn get_mut_nochange(&mut self, coord: &Point3<u32>) -> Option<&mut T>;
+
+    /// Try to get the `Tile` type present at the provided coordinates.
+    fn get_raw(&self, coord: u32) -> Option<&T>;
+
+    /// Try to get a mutable reference to `Tile` type present at the provided coordinates.
+    fn get_raw_mut(&mut self, coord: u32) -> Option<&mut T>;
+
+    /// Try to get a mutable reference to `Tile` type present at the provided coordinates.
+    fn get_raw_mut_nochange(&mut self, coord: u32) -> Option<&mut T>;
+}
+
+/// Concrete implementation of a generic 3D `TileMap` component. Accepts a `Tile` type and `CoordinateEncoder` type,
+/// creating a flat 1D array storage which is spatially partitioned utilizing the provided encoding scheme.
+/// creating a flat 1D array storage which is spatially partitioned utilizing the provided encoding scheme.
+///
+/// The defualt encoding scheme is `MortonEncoder2D`, which allows for arbitrary X, Y and Z coordinate sizes while
+/// still spatially partitioning each z-level. For more efficient Z-order encoding, use `MortonEncoder` which requires
+/// cubic map dimensions but provides for much greater spatial efficiency.
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct TileMap<T: Tile, E: CoordinateEncoder = crate::MortonEncoder2D> {
+    pub(crate) origin: Point3<f32>,
+    pub(crate) tile_dimensions: Vector3<u32>,
+    pub(crate) dimensions: Vector3<u32>,
+    pub(crate) transform: Matrix4<f32>,
+
+    pub(crate) version: u64,
+
+    #[serde(skip)]
+    pub(crate) sprite_sheet: Option<Handle<SpriteSheet>>,
```

Comment:
If it's skipped for serialization, how is it supposed to be bound after?

Reply:
I don't know. I didn't design the `Prefab` and `Asset` system, its crap. How is this supposed to happen?

Reply:
Done.